### ISSUE TITLE
Changes to GridTool Cache

### ISSUE THREAD

**Issue by GivAlz** [dealii/dealii#7536]
Hi everyone,
  I am posting this PR because, in the future, I hope to improve the functionality of distributed compute point location, but I have a design problem to solve (and time to find) first.

Once PR https://github.com/dealii/dealii/pull/7001 is closed, we shall have a reliable instrument to build a description of the whole mesh using bounding boxes (call it "global rtree" ): using the local rtree of bounding boxes,call rtree->bounds() function which returns a single box surrounding all local cells (this works, even though it is not always optimal), and then call the function to build a global bounding boxes tree. Then, some simple modifications to distributed compute point location should make the function faster and easier to read...

The process I described can be easily implemented, but it would be nice to automate it. The naive approach would be to add the "global rtree" to the Cache, and add somewhere a function which generates and stores the "global rtree" in the Cache. The main problem with this approach is that this involves global communication (sigh!) and Cache won't be able to update the global tree without user intervention (or risking a communication deadlock).

My question is: would it be better to leave everything as a function outside Cache, and let the user handle the construction of a global tree "by hand", or would it be "acceptable" to add this feature to the Cache with the warning that, for this particular class member, the user is the only one who can (and has to) update it before using it? What do you think @luca-heltai ?

**Comment by luca-heltai** [comment]
Notice that a global bounding box tree is already in grid tools cache, and can be used by the following **super simple** find active cell around point:

```

#include <deal.II/base/patterns.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_out.h>
#include <deal.II/grid/grid_tools.h>
#include <deal.II/grid/grid_tools_cache.h>
#include <deal.II/grid/tria.h>
#include <deal.II/grid/tria_accessor.h>
#include <deal.II/grid/tria_iterator.h>

#include <deal.II/numerics/rtree.h>

#include <cmath>
#include <fstream>
#include <iostream>
#include <unordered_map>

#include "../tests.h"

template <int dim, int spacedim>
struct cell_hash
{
  std::size_t operator()(
    const typename Triangulation<dim, spacedim>::active_cell_iterator &k) const
  {
    // Return active cell index, which is faster than CellId to compute
    return k->active_cell_index();
  }
};


namespace bgi = boost::geometry::index;

template <int dim, int spacedim>
std::pair<std::unordered_multimap<
            typename Triangulation<dim, spacedim>::active_cell_iterator,
            std::pair<Point<dim>, unsigned int>,
            cell_hash<dim, spacedim>>,
          std::vector<unsigned int>>
compute_point_locations_rtree(const GridTools::Cache<dim, spacedim> &cache,
                              const std::vector<Point<spacedim>> &   points,
                              const double &tol = 1e-12)
{
  const auto &mapping = cache.get_mapping();
  const auto &c_tree  = cache.get_cell_bounding_boxes_rtree();

  // map point index to intersecting boxes
  std::vector<std::vector<
    std::pair<BoundingBox<spacedim>,
              typename Triangulation<dim, spacedim>::active_cell_iterator>>>
    boxes(points.size());

  // Track the index of the point p
  unsigned int i = 0;
  for (const auto &p : points)
    c_tree.query(bgi::intersects(p), std::back_inserter(boxes[i++]));

  std::unordered_multimap<
    typename Triangulation<dim, spacedim>::active_cell_iterator,
    std::pair<Point<dim>, unsigned int>,
    cell_hash<dim, spacedim>>
    mapped_points;

  std::vector<unsigned int> outside_points;

  // Create a unit bounding box, with boundary extended according to the
  // input tolerance value
  BoundingBox<dim> unit_box;
  {
    Point<dim> p0, p1;
    for (unsigned int d = 0; d < dim; ++d)
      {
        p0[d] = -tol;
        p1[d] = 1 + tol;
      }
    unit_box = BoundingBox<dim>({p0, p1});
  }

  for (unsigned int i = 0; i < points.size(); ++i)
    {
      bool point_found = false;
      for (unsigned int j = 0; j < boxes[i].size(); ++j)
        {
          const auto &box  = boxes[i][j];
          const auto &cell = box.second;
          const auto  p_ref =
            mapping.transform_real_to_unit_cell(cell, points[i]);
          if (unit_box.point_inside(p_ref))
            {
              // We found it inside one of the bounding boxes
              mapped_points.insert(
                std::make_pair(cell, std::make_pair(p_ref, i)));
              point_found = true;
              break;
            }
        }
      if (point_found == false)
        outside_points.push_back(i);
    }
  return {mapped_points, outside_points};
}

template <int dim, int spacedim>
std::tuple<
  std::vector<typename Triangulation<dim, spacedim>::active_cell_iterator>,
  std::vector<std::vector<Point<dim>>>,
  std::vector<std::vector<unsigned int>>>
map_to_tuple(const std::unordered_multimap<
             typename Triangulation<dim, spacedim>::active_cell_iterator,
             std::pair<Point<dim>, unsigned int>,
             cell_hash<dim, spacedim>> &map_of_points)
{
  std::vector<typename Triangulation<dim, spacedim>::active_cell_iterator>
                                         cells;
  std::vector<std::vector<Point<dim>>>   points;
  std::vector<std::vector<unsigned int>> indices;

  for (unsigned int b = 0; b < map_of_points.bucket_count(); ++b)
    {
      auto n = map_of_points.bucket_size(b);
      if (n > 0)
        {
          auto it = map_of_points.begin(b);
          cells.push_back(it->first);
          points.push_back(std::vector<Point<dim>>(n));
          indices.push_back(std::vector<unsigned int>(n));

          for (unsigned int j = 0; it != map_of_points.end(b); ++it, ++j)
            {
              const auto &p     = it->second.first;
              const auto &id    = it->second.second;
              points.back()[j]  = p;
              indices.back()[j] = id;
            }
        }
    }
  return {cells, points, indices};
}


int main()
{
  initlog(1);
  const unsigned int dim = 2, spacedim = 2;
  const unsigned int N = 100;

  Triangulation<dim, spacedim> triangulation;

  GridGenerator::hyper_ball(triangulation, Point<spacedim>(), 2);
  triangulation.refine_global(2);

  GridTools::Cache<dim, spacedim> cache(triangulation);

  std::vector<Point<spacedim>> points(N);
  for (auto &p : points)
    p = random_point<spacedim>();

  auto [res, outside] = compute_point_locations_rtree(cache, points);

  auto [cells, ref_points, indices] = map_to_tuple(res);

  deallog << "Cells with points inside: " << cells.size() << "/"
          << triangulation.n_active_cells() << std::endl;
  deallog << "Missed points           : " << outside.size() << "/" << N
          << std::endl;
}
```

**Comment by luca-heltai** [comment]
I think people have different meaning w.r.t. what a "global bounding box tree of the triangulation" is. What is generally meant (and what is available in `Cache`) is a tree constructed with **all** cell bounding boxes. 

This object can (and should) be queried for fast (and possibly complex) intersection queries. What you call a "global bounding box tree" is an object constructed by hand by one the functions that you implemented that represents the minimal faithful covering of the triangulation (to guess who may own an arbitrary point in space). 

These two should not be confused with each other, so I suggest we pick up a naming convention that helps distinguish things. Maybe `minimal_covering`, rather than `global_bounding_box`.

It's unclear to me if there is a default way to construct this, and discussions with @tjhei suggested (back in July) that there is *not*. However, once we agree on *what* to store in the Cache, I don't think there is a deadlock problem: 

The Cache is invalidated by any change to the Triangulation, and whenever you query for the minimal covering, if the Cache has been invalidated, it can be reconstructed (once) and passed around (once) to be available to anybody needing it.

**Comment by GivAlz** [comment]
Ok... I like your proposal of "minimal covering" (or maybe triangulation covering? ), the function was implemented with a different name (sorry!). But I guess it can be changed and then stored in the cache, this shouldn't be difficult. Thank you for the tip! :)

**Comment by drwells** [comment]
I believe this has since been fixed (especially by #7583).
